src/contract/contract.cpp

```diff
@@ -0,0 +1,122 @@
+    if(!bValid)
+        LogPrintf("VerifyCPIDSignature: invalid signature sSignature=%s, cached key=%s\n"
+                  ,sSignature, sBeaconPublicKey);
+    return bValid;
+}
+
+std::string SignMessage(std::string sMsg, std::string sPrivateKey)
+{
+    CKey key;
+    std::vector<unsigned char> vchMsg = std::vector<unsigned char>(sMsg.begin(), sMsg.end());
+    std::vector<unsigned char> vchPrivKey = ParseHex(sPrivateKey);
+    std::vector<unsigned char> vchSig;
+    key.SetPrivKey(CPrivKey(vchPrivKey.begin(), vchPrivKey.end())); // if key is not correct openssl may crash
+    if (!key.Sign(Hash(vchMsg.begin(), vchMsg.end()), vchSig))
+    {
+        return "Unable to sign message, check private key.";
+    }
+    const std::string sig(vchSig.begin(), vchSig.end());
+    std::string SignedMessage = EncodeBase64(sig);
+    return SignedMessage;
+}
+
+std::string AddMessage(bool bAdd, std::string sType, std::string sPrimaryKey, std::string sValue,
+                       std::string sMasterKey, int64_t MinimumBalance, double dFees, std::string strPublicKey)
+{
+    std::string sAddress = GetBurnAddress();
+    CBitcoinAddress address(sAddress);
+    if (!address.IsValid())       throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid Gridcoin address");
+    int64_t nAmount = AmountFromValue(dFees);
+    // Wallet comments
+    CWalletTx wtx;
+    if (pwalletMain->IsLocked())  throw JSONRPCError(RPC_WALLET_UNLOCK_NEEDED, "Error: Please enter the wallet passphrase with walletpassphrase first.");
+    std::string sMessageType      = "<MT>" + sType  + "</MT>";  //Project or Smart Contract
+    std::string sMessageKey       = "<MK>" + sPrimaryKey   + "</MK>";
+    std::string sMessageValue     = "<MV>" + sValue + "</MV>";
+    std::string sMessagePublicKey = "<MPK>"+ strPublicKey + "</MPK>";
+    std::string sMessageAction    = bAdd ? "<MA>A</MA>" : "<MA>D</MA>"; //Add or Delete
+    //Sign Message
+    std::string sSig = SignMessage(sType+sPrimaryKey+sValue,sMasterKey);
+    std::string sMessageSignature = "<MS>" + sSig + "</MS>";
+    wtx.hashBoinc = sMessageType+sMessageKey+sMessageValue+sMessageAction+sMessagePublicKey+sMessageSignature;
+    std::string strError = pwalletMain->SendMoneyToDestinationWithMinimumBalance(address.Get(), nAmount, MinimumBalance, wtx);
+    if (!strError.empty())        throw JSONRPCError(RPC_WALLET_ERROR, strError);
+    return wtx.GetHash().GetHex().c_str();
+}
+
+std::string AddContract(std::string sType, std::string sName, std::string sContract)
+{
+    std::string sPass = (sType=="project" || sType=="projectmapping" || sType=="smart_contract") ? GetArgument("masterprojectkey", msMasterMessagePrivateKey) : msMasterMessagePrivateKey;
+    std::string result = AddMessage(true,sType,sName,sContract,sPass,AmountFromValue(1),.00001,"");
+    return result;
+}
+
+bool SignBlockWithCPID(const std::string& sCPID, const std::string& sBlockHash, std::string& sSignature, std::string& sError, bool bAdvertising=false)
```

Comment:
This function is more releated to beacon, or miner, than contracts.

Reply:
Is it fine if I leave it there for now, and group it back, when I get to moving the beacon and miner code as well?

Reply:
iirc there is already a beacon.cpp file, but I guess it is OK.